src/lib/components/Table/Table.js

```diff
@@ -0,0 +1,199 @@
+    }
+  }
+
+  generateData() {
+    const { children } = this.props;
+    const columns = [];
+    const data = [];
+
+    children.forEach((row, rowIndex) => {
+      if (row.props.tableHeader) {
+        row.props.children.forEach((cell, cellIndex) => {
+          let value = '';
+          if (typeof cell.props.children === 'string') value = cell.props.children;
+          columns.push({
+            key: `column${cellIndex}`,
+            label: value,
+            align: cell.props.align,
+            header: cell.props.columnHeader,
+          });
+        });
+      } else {
+        const index = rowIndex - 1;
+        data[index] = {};
+        data[index].id = rowIndex;
+        row.props.children.forEach((cell, cellIndex) => {
+          let value = '';
+          if (typeof cell.props.children === 'string') value = cell.props.children;
+          data[index][`column${cellIndex}`] = value;
+        });
+      }
+    });
+
+    this.setState({
+      columns,
+      data: [...data],
+      originalData: data,
+    });
+  }
+
+  generateHeader() {
+    const { columns, sortCondition } = this.state;
+    const { sortable } = this.props;
+
+    const columnHeaders = columns.map((headerData) => {
+      const ariaSort = sortCondition[headerData.key];
+      const onClick = sortable ? () => this.getNewSortCondition(headerData.key) : null;
+      return (
+        <TableCell
+          columnHeader
+          key={headerData.key}
+          ariaSort={ariaSort}
+          align={headerData.align}
+          onClick={onClick}
+        >
+          {headerData.label}
+        </TableCell>);
+    });
+
+    const header = (
+      <thead>
+        <TableRow key="header-row">
+          {columnHeaders}
+        </TableRow>
+      </thead>
+    );
+
+    return header;
+  }
+
+  generateRows() {
+    const { columns, data } = this.state;
+    const rows = data.map((item) => {
+      const cells = columns.map(headerData => (
+        <TableCell
+          rowHeader={headerData.header}
+          key={`${headerData.key}-row-${item.id}`}
+          align={headerData.align}
+        >
+          {item[headerData.key]}
+        </TableCell>),
+      );
+      return (
+        <TableRow key={`row-${item.id}`}>
+          {cells}
+        </TableRow>
+      );
+    });
+
+    return rows;
+  }
+
+  render() {
+    const { hasHeader, sortable } = this.props;
+    const className = sortable ? 'p-table--sortable' : null;
+
+    return (
+      <table role="grid" className={className}>
+        {hasHeader && this.generateHeader()}
+        <tbody>
+          {this.generateRows()}
+        </tbody>
+      </table>
+    );
+  }
+}
+
+Table.defaultProps = {
+  children: null,
+  columns: null,
+  data: null,
+  hasHeader: true,
+  sortable: false,
+  sortCondition: {},
+};
+
+Table.propTypes = {
+  children: PropTypes.node,
+  columns: PropTypes.arrayOf(PropTypes.object),
+  data: PropTypes.arrayOf(PropTypes.object),
+  hasHeader: PropTypes.bool,
```

Comment:
Is this not defined by the data provided? Or would we want to override to hide a header? Would we not just omit the header from the data? So many questions
![So many questions](https://media.giphy.com/media/5XRB3Ay93FZw4/giphy.gif)

Reply:
This is a weird one. Basically you neeeeeeed a columns prop to provide a reference for how the data prop is mapped into TableCells (not sure how you'd do it without a data AND columns prop). And it's currently set up so that a header is automatically generated from the columns prop. So yeah it's basically there to override the header automatically showing